reactivesocket-client/src/main/java/io/reactivesocket/client/LoadBalancer.java

```diff
@@ -618,4 +661,284 @@ public void shutdown() {}
+     * update availability accordingly.
+     */
+    private class WeightedSocket extends ReactiveSocketProxy {
+        private static final double STARTUP_PENALTY = Long.MAX_VALUE >> 12;
+
+        private final ReactiveSocket child;
+        private ReactiveSocketFactory<?> factory;
+        private final Quantile lowerQuantile;
+        private final Quantile higherQuantile;
+        private final long inactivityFactor;
+
+        private volatile int pending;       // instantaneous rate
+        private long stamp;                 // last timestamp we sent a request
+        private long stamp0;                // last timestamp we sent a request or receive a response
+        private long duration;              // instantaneous cumulative duration
+
+        private Median median;
+        private Ewma interArrivalTime;
+
+        private AtomicLong pendingStreams;  // number of active streams
+
+        WeightedSocket(
+            ReactiveSocket child,
+            ReactiveSocketFactory<?> factory,
+            Quantile lowerQuantile,
+            Quantile higherQuantile,
+            int inactivityFactor
+        ) {
+            super(child);
+            this.child = child;
+            this.factory = factory;
+            this.lowerQuantile = lowerQuantile;
+            this.higherQuantile = higherQuantile;
+            this.inactivityFactor = inactivityFactor;
+            long now = Clock.now();
+            this.stamp = now;
+            this.stamp0 = now;
+            this.duration = 0L;
+            this.pending = 0;
+            this.median = new Median();
+            this.interArrivalTime = new Ewma(1, TimeUnit.MINUTES, 1000);
+            this.pendingStreams = new AtomicLong();
+        }
+
+        WeightedSocket(
+            ReactiveSocket child,
+            ReactiveSocketFactory<?> factory,
+            Quantile lowerQuantile,
+            Quantile higherQuantile
+        ) {
+            this(child, factory, lowerQuantile, higherQuantile, 100);
+        }
+
+        @Override
+        public Publisher<Payload> requestResponse(Payload payload) {
+            return subscriber ->
+                child.requestResponse(payload).subscribe(new LatencySubscriber<>(subscriber, this));
+        }
+
+        @Override
+        public Publisher<Payload> requestStream(Payload payload) {
+            return subscriber ->
+                child.requestStream(payload).subscribe(new CountingSubscriber<>(subscriber, this));
+        }
+
+        @Override
+        public Publisher<Payload> requestSubscription(Payload payload) {
+            return subscriber ->
+                child.requestSubscription(payload).subscribe(new CountingSubscriber<>(subscriber, this));
+        }
+
+        @Override
+        public Publisher<Void> fireAndForget(Payload payload) {
+            return subscriber ->
+                child.fireAndForget(payload).subscribe(new CountingSubscriber<>(subscriber, this));
```

Comment:
This will include write latency to the socket, which _may_ be useful as we do not currently listen to `keep-alive` latency? Probably more if this is the primary interaction model for a client.

I think you said that the only _really_ supported model is request-response, so this may just be moot.